# Machine Learning Zoomcamp 2026 - Homework 1: Introduction to Machine Learning

Este notebook contiene la solución completa, explicada y reproducible de la Tarea 1 de la cohorte 2026.

### Objetivos:
- Configuración y verificación del entorno.
- Análisis exploratorio con Pandas y NumPy.
- Cálculo de estadísticos básicos (media, mediana, moda).
- Implementación de regresión lineal en su formulación matricial normal ($w = (X^T X)^{-1} X^T y$).


## 1. Importación de Librerías y Configuración

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path

print(f"NumPy version: {np.__version__}")
print(f"Pandas version: {pd.__version__}")


NumPy version: 2.4.6
Pandas version: 3.0.6


## Q1. Pandas version
¿Qué versión de Pandas tienes instalada?

In [3]:
# Pregunta 1
pd.__version__


'3.0.6'

## Carga del Dataset
Usamos el dataset fijado para la cohorte 2026: `car_fuel_efficiency_2026.csv`.

In [4]:
data_path = Path("../../data/car_fuel_efficiency_2026.csv")
df = pd.read_csv(data_path)
df.head()


,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


## Q2. Records count
¿Cuántos registros tiene el dataset?
- 5000
- 9000
- 10000
- 15000

In [5]:
# Pregunta 2: Número de registros
n_records = len(df)
print(f"Registros en el dataset: {n_records}")


Registros en el dataset: 10000


## Q3. Fuel types
¿Cuántos tipos de combustible (`fuel_type`) están presentes en el dataset?
- 1
- 2
- 3
- 4

In [6]:
# Pregunta 3: Tipos de combustible
fuel_types = df["fuel_type"].unique()
n_fuel_types = df["fuel_type"].nunique()
print(f"Tipos encontrados ({n_fuel_types}): {fuel_types.tolist()}")


Tipos encontrados (3): ['Gasoline', 'Diesel', 'Hybrid']


## Q4. Missing values
¿Cuántas columnas tienen valores faltantes en el dataset?
- 0
- 1
- 2
- 3
- 4

In [7]:
# Pregunta 4: Columnas con valores nulos
missing_counts = df.isna().sum()
missing_cols = missing_counts[missing_counts > 0]
print("Columnas con nulos:")
print(missing_cols)
print(f"Total de columnas con valores faltantes: {len(missing_cols)}")


Columnas con nulos:
horsepower      877
acceleration    264
dtype: int64
Total de columnas con valores faltantes: 2


## Q5. Max fuel efficiency
¿Cuál es la máxima eficiencia de combustible (`fuel_efficiency_mpg`) de los autos de origen Asia?
- 21.2
- 31.2
- 41.2
- 51.2

In [8]:
# Pregunta 5: Máximo fuel efficiency para autos de Asia
asia_cars = df[df["origin"] == "Asia"]
max_eff_asia = asia_cars["fuel_efficiency_mpg"].max()
print(f"Máxima eficiencia en Asia: {max_eff_asia:.1f}")


Máxima eficiencia en Asia: 41.2


## Q6. Median value of horsepower
1. Hallar la mediana de `horsepower`.
2. Calcular el valor más frecuente (moda) de `horsepower`.
3. Rellenar los valores nulos con la moda.
4. Recalcular la mediana de `horsepower`.

¿Ha cambiado? (`Yes, it increased`, `Yes, it decreased`, `No`)

In [9]:
# Pregunta 6: Mediana antes y después de imputar moda
horsepower_col = df["horsepower"]
median_before = horsepower_col.median()
mode_hp = horsepower_col.mode().iloc[0]

horsepower_imputed = horsepower_col.fillna(mode_hp)
median_after = horsepower_imputed.median()

print(f"Mediana inicial: {median_before}")
print(f"Moda imputada:   {mode_hp}")
print(f"Mediana final:   {median_after}")

if median_after < median_before:
    res_q6 = "Yes, it decreased"
elif median_after > median_before:
    res_q6 = "Yes, it increased"
else:
    res_q6 = "No"

print(f"Resultado: {res_q6}")


Mediana inicial: 254.0
Moda imputada:   252.0
Mediana final:   252.0
Resultado: Yes, it decreased


## Q7. Sum of weights (Linear Regression)
1. Filtrar autos de Asia.
2. Seleccionar columnas `vehicle_weight` y `model_year`.
3. Tomar los primeros 7 valores (`X`).
4. Calcular $X^T X$.
5. Invertir $X^T X$.
6. Crear $y = [1100, 1300, 800, 900, 1000, 1100, 1200]$.
7. Calcular $w = (X^T X)^{-1} X^T y$.
8. ¿Cuál es la suma de los elementos de $w$?

In [10]:
# Pregunta 7: Implementación normal de regresión lineal
asia_subset = df.loc[df["origin"] == "Asia", ["vehicle_weight", "model_year"]].head(7)
X = asia_subset.to_numpy()
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

XTX = X.T @ X
XTX_inv = np.linalg.inv(XTX)
w = XTX_inv @ X.T @ y

print(f"Matriz X (7x2):\n{X}")
print(f"Vector y: {y}")
print(f"Pesos w: {w}")
print(f"Suma de elementos de w: {w.sum():.3f}")


Matriz X (7x2):
[[4240 1996]
 [4310 1992]
 [4230 1997]
 [4420 1978]
 [3950 2007]
 [4750 2021]
 [4450 2019]]
Vector y: [1100 1300  800  900 1000 1100 1200]
Pesos w: [0.13644777 0.2327492 ]
Suma de elementos de w: 0.369


### Resumen de Respuestas HW1:
- **Q1**: Ver `pd.__version__`
- **Q2**: `10000`
- **Q3**: `3`
- **Q4**: `2`
- **Q5**: `41.2`
- **Q6**: `Yes, it decreased`
- **Q7**: `0.369`
